nni/contrib/compression/pruning/basic_pruner.py

```diff
@@ -0,0 +1,194 @@
+
+    def _collect_data(self) -> _DATA:
+        return active_sparse_targets_filter(self._target_spaces)
+
+    def _calculate_metrics(self, data: _DATA) -> _METRICS:
+        return norm_metrics(p=self.p, data=data, target_spaces=self._target_spaces)
+
+    def _generate_sparsity(self, metrics: _METRICS) -> _MASKS:
+        return generate_sparsity(metrics, self._target_spaces)
+
+    def _register_trigger(self, evaluator: Evaluator):
+        self._current_step = 0
+        self._is_first_step = True
+        self._remaining_times = self.total_times
+
+        def optimizer_task():
+            self._current_step += 1
+            if (self._is_first_step and self.first_step_gen) or self._current_step == self.interval_steps:
+                masks = self.generate_masks()
+                self.update_masks(masks)
+                if isinstance(self._remaining_times, int):
+                    self._remaining_times -= 1
+                debug_msg = f'{self.__class__.__name__} generate masks, remaining times {self._remaining_times}'
+                _logger.debug(debug_msg)
+            if self._current_step == self.interval_steps and \
+                (self._remaining_times == 'unlimited' or self._remaining_times > 0):  # type: ignore
+                self._current_step = 0
+            if self._is_first_step:
+                self._is_first_step = False
+
+        evaluator.patch_optimizer_step(before_step_tasks=[], after_step_tasks=[optimizer_task])
+
+    def _single_compress(self, max_steps: None, max_epochs: None):
+        assert max_steps is None and max_epochs is None, \
+            f'{self.__class__.__name__} do not support training aware pruning under single compress mode.'
+        masks = self.generate_masks()
+        self.update_masks(masks)
+
+    def _fuse_preprocess(self, evaluator: Evaluator):
+        self._register_trigger(evaluator)
+
+    def _fuse_postprocess(self, evaluator: Evaluator):
+        pass
+
+    @overload
+    def compress(self) -> Tuple[torch.nn.Module, _MASKS]:
+        ...
+
+    @overload
+    def compress(self, max_steps: int | None, max_epochs: int | None) -> Tuple[torch.nn.Module, _MASKS]:
+        ...
+
+    def compress(self, max_steps: int | None = None, max_epochs: int | None = None):
+        return super().compress(max_steps, max_epochs)
+
+
+class LevelPruner(_NormPruner):
+    """
+    This is a basic pruner, and in some papers called it magnitude pruning or fine-grained pruning.
+    It will mask the smallest magnitude weights in each specified layer by a saprsity ratio configured in the config list.
+
+    Parameters
+    ----------
+    model
+        Model to be pruned.
+    config_list
+        A list of dict, each dict configure which module need to be pruned, and how to prune.
+        Please refer :doc:`Compression Config Specification </compression/compression_config_list>` for more information.
+
+    Examples
+    --------
+        TODO
```

Comment:
TODO? why not add it in this pr?

Reply:
a bit large, need add multi example files in this pr, will add them with tests in another pr (on going...)